projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,313 @@
+Lib->>HSM: Ensure key exists
+HSM-->>Lib: Confirm Key exists
+Lib->>HSM: Generate the `xPub`
+HSM->>Lib: Return the `xPub`
+Note right of Lib: Initialize hardware wallet
+Lib->>Headless: Wallet initialized
+Note right of Headless: Regular wallet usage
+Headless->>Lib: Request to sign an input for<br/>Tx Y, address X
+Lib->>HSM: Sign string with<br/>Child Key derivation on path X
+HSM->>Lib: Signed string
+Lib->>Lib: Assemble Tx
+Lib->>Headless: Tx Y with signed input
+```
+
+Eventual network failures when trying to communicate with the HSM are retried and only interrupt the flow if the amount of retries reach a configurable threshold. Logic errors like not finding an `xPriv` key are treated and returned to the end user in an informative way.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This project has three main phases that need implementing:
+1. The communication with the HSM itself. We will call this the *HSM Client* throughout this document.
+2. Adapting the Headless wallet to interact with this client and use the HSM for cryptographic operations
+
+## The HSM Client
+The main library offered by Dinamo to interact with the HSM device [is available in C language](https://manual.dinamonetworks.io/c/index.html).
+
+A proof-of concept application was partially developed in C as a bridge between the Wallet Lib javascript environment and the C environment. A dedicated design will be written about this interface, but in short:
+- The C code expects to be called from the operating system. In Node.js terms this means using [`child_process.spawn()`](https://nodejs.org/docs/latest-v18.x/api/child_process.html#child_processspawncommand-args-options)
+- Parameters such as the derivation path and input data to be signed should be passed through `stdin`, not the `argv` command parameters. This avoids leaking sensitive data to the public command line history of the OS
+- Responses are sent through `stdout` and read using the child process `stdout.on('data')` event listener
+
+This means that headless applications that are supposed to interact with the HSM should copy the compiled executables of the *HSM Client* to their environment before or during operation.
+
+## Headless implementation
+[headless-implementation]: #headless-implementation
+The headless application will be responsible for calling the *HSM Client* directly.
+
+The connection authentication credentials will be stored securely on the environment variables, allowing the leverage of other security best practices to avoid credential leaks.
+```sh
+# On env variables
+HSM_DEVICE_HOST="192.168.0.1"
+HSM_DEVICE_USERNAME="my_username"
+HSM_DEVICE_PASSWORD="my_password"
+```
+
+A configuration attribute will be added to the `config.js` file, similar to the existing `seeds` attribute, mapping the headless `walletId`s to HSM object key names.
+```js
+// On config.js
+const hsmWallets = {
+  'headless-wallet-id': 'HSM_KEY_NAME'
+}
+```
```

Comment:
When running on docker (the intended way for headless) the config.js loads the configuration from the environment variables, so we should also show how the environment variables fot this config will be done.

Another point to be sure is that we have a system to reload configuration with the wallet running.
We need to map on the settings module how the wallet will react if this configuration changes.

Reply:
Good point! The description was simplified to accomodate the new Settings module and soft reload feature.